pallets/thea/src/lib.rs

```diff
@@ -0,0 +1,534 @@
+
+				if !deposits.is_empty() {
+					// If pending deposits are available, save it back
+					<ApprovedDeposits<T>>::insert(&user, deposits)
+				} else {
+					<AccountWithPendingDeposits<T>>::mutate(|accounts| accounts.remove(&user));
+				}
+			}
+
+			Ok(())
+		}
+
+		/// Extrinsic to update withdrawal completion status by relayer
+		///
+		/// # Parameters
+		///
+		/// * `origin`: Any relayer
+		/// * `withdrawal_nonce`: Withdrawal Nonce
+		/// * `network`: Network id
+		/// * `tx_hash`: Vec<u8>
+		/// * `bit_map`: Bitmap of Thea relayers
+		/// * `bls_signature`: BLS signature of relayers
+		#[pallet::weight(1000)]
+		pub fn batch_withdrawal_complete(
+			origin: OriginFor<T>,
+			withdrawal_nonce: u32,
+			network: Network,
+			tx_hash: sp_core::H256,
+			_bit_map: u128,
+			_bls_signature: [u8; 96],
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			// TODO: Something to verify BLS signature
+			<ReadyWithdrawls<T>>::take(network, withdrawal_nonce);
+
+			Self::deposit_event(Event::<T>::WithdrawalExecuted(withdrawal_nonce, network, tx_hash));
+			Ok(())
+		}
+
+		/// Initiate the withdraw for user
+		///
+		/// # Parameters
+		///
+		/// * `origin`: User
+		/// * `asset_id`: Asset id
+		/// * `amount`: Amount of asset to withdraw
+		#[pallet::weight(1000)]
+		pub fn withdraw(
+			origin: OriginFor<T>,
+			asset_id: u128,
+			amount: u128,
+			beneficiary: Vec<u8>,
+			pay_for_remaining: bool,
+		) -> DispatchResult {
+			let user = ensure_signed(origin)?;
+			// Put a soft limit of size of beneficiary vector to avoid spam
+			ensure!(beneficiary.len() <= 100, Error::<T>::BeneficiaryTooLong);
+
+			// Find native network of this asset
+			#[allow(clippy::unnecessary_lazy_evaluations)]
+			// TODO: Remove once merged to asset-handler
+			let network = <AssetIdToNetworkMapping<T>>::get(asset_id)
+				.ok_or_else(|| Error::<T>::UnableFindNetworkForAssetId)?;
+
+			let withdrawal_nonce = <WithdrawalNonces<T>>::get(network);
+
+			let mut pending_withdrawals = <PendingWithdrawals<T>>::get(network);
+
+			// Ensure pending withdrawals have space for a new withdrawal
+			ensure!(pending_withdrawals.is_full(), Error::<T>::WithdrawalNotAllowed);
+
+			#[allow(clippy::unnecessary_lazy_evaluations)]
+			// TODO: Remove once merged to asset-handler
+			let mut total_fees = <WithdrawalFees<T>>::get(network)
+				.ok_or_else(|| Error::<T>::WithdrawalFeeConfigNotFound)?;
+
+			if pay_for_remaining {
+				// User is ready to pay for remaining pending withdrawal for quick withdrawal
+				let extra_withdrawals_available = 10usize.saturating_sub(pending_withdrawals.len());
+				total_fees = total_fees.saturating_add(
+					total_fees.saturating_mul(extra_withdrawals_available.saturated_into()),
+				)
+			}
+
+			// Pay the fees
+			<T as Config>::Currency::transfer(
+				&user,
+				&Self::thea_account(),
+				total_fees.saturated_into(),
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			// TODO: Update Thea Staking pallet about fees collected
```

Comment:
Please add some task reference to not forget

Reply:
Done